src/core/lib/math/distributiongenerator.cpp

```diff
@@ -35,12 +35,142 @@
  */
 
 #include "math/distributiongenerator.h"
+#include "utils/prng/blake2engine.h"
+#include "utils/exception.h"
 
-#include <memory>
+#include <chrono>
+#include <cstdint>
+#include <dlfcn.h>
 #include <random>
+#include <thread>
+// #include <iostream>
 
 namespace lbcrypto {
 
 std::shared_ptr<PRNG> PseudoRandomNumberGenerator::m_prng = nullptr;
+PseudoRandomNumberGenerator::GenPRNGEngineFuncPtr PseudoRandomNumberGenerator::genPRNGEngine = nullptr;
+
+void PseudoRandomNumberGenerator::InitPRNGEngine(const std::string& libPath) {
+    if(genPRNGEngine) // if genPRNGEngine has already been initialized
+      return;
+
+    if(libPath.empty()) {
+        // use the default OpenFHE PRNG that comes with the library 
+        genPRNGEngine = default_prng::createEngineInstance;
+        if (!genPRNGEngine)
+            OPENFHE_THROW("Cannot find symbol: default_prng::createEngineInstance");
+        // std::cerr << "InitPRNGEngine: using local PRNG" << std::endl;
+    }
+    else {
+        // do not close libraryHandle, your application will crash if you do
+        void* libraryHandle = dlopen(libPath.c_str(), RTLD_LAZY);
+        if (!libraryHandle) {
+            std::string errMsg{std::string("Cannot open ") + libPath + ": "};
+            const char* dlsym_error = dlerror();
+            errMsg += dlsym_error;
+            OPENFHE_THROW(errMsg);
+        }
+        genPRNGEngine = (GenPRNGEngineFuncPtr)dlsym(libraryHandle, "createEngineInstance");
+        if (!genPRNGEngine) {
+            std::string errMsg{std::string("Cannot load symbol createEngineInstance() from ") + libPath};
+            const char* dlsym_error = dlerror();
+            errMsg += ": ";
+            errMsg += dlsym_error;
+            dlclose(libraryHandle);
+            OPENFHE_THROW(errMsg);
+        }
+        // std::cerr << "InitPRNGEngine: using external PRNG" << std::endl;
+    }
+} 
+
+PRNG& PseudoRandomNumberGenerator::GetPRNG() {
+    // initialization of PRNGs
+    if (m_prng == nullptr) {
+#pragma omp critical
+        {
+            if (!genPRNGEngine)
+                InitPRNGEngine();
+
```

Comment:
Should there be a check here that genPRNGEngine is no longer nullptr?

I realize this check is perfomed in the called code.  I also realize that it is declared static, so not able to be overridden.

This may be a no-op comment.